src/compiler/checker.ts

```diff
@@ -6638,22 +6638,31 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 const questionToken = type.declaration.questionToken ? factory.createToken(type.declaration.questionToken.kind) as QuestionToken | PlusToken | MinusToken : undefined;
                 let appropriateConstraintTypeNode: TypeNode;
                 let newTypeVariable: TypeReferenceNode | undefined;
+                let templateType = getTemplateTypeFromMappedType(type);
+                let typeParameter = getTypeParameterFromMappedType(type);
                 if (isMappedTypeWithKeyofConstraintDeclaration(type)) {
                     // We have a { [P in keyof T]: X }
                     // We do this to ensure we retain the toplevel keyof-ness of the type which may be lost due to keyof distribution during `getConstraintTypeFromMappedType`
                     if (isHomomorphicMappedTypeWithNonHomomorphicInstantiation(type) && context.flags & NodeBuilderFlags.GenerateNamesForShadowedTypeParams) {
-                        const newParam = createTypeParameter(createSymbol(SymbolFlags.TypeParameter, "T" as __String));
-                        const name = typeParameterToName(newParam, context);
+                        const newConstraintParam = createTypeParameter(createSymbol(SymbolFlags.TypeParameter, "T" as __String));
+                        const newTypeParam = createTypeParameter(createSymbol(SymbolFlags.TypeParameter, "K" as __String));

```

Comment:
Also, it would definitely be better just to reuse the existing mapped type's key type parameter wholesale; I can't really think of why we'd need a unique copy here, and it's usually better to keep type identities minimal where possible.

Reply:
I've applied the first suggestion (reusing the symbol). I'm not *exactly* sure what's the expected result that I'm supposed to get here but I think I might be somewhat close using this extra diff:
<details>
<summary>git diff</summary>

```diff
diff --git a/src/compiler/checker.ts b/src/compiler/checker.ts
index dfbcd01413..33bbc22259 100644
--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -6642,7 +6642,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 let appropriateConstraintTypeNode: TypeNode;
                 let newTypeVariable: TypeReferenceNode | undefined;
                 let templateType = getTemplateTypeFromMappedType(type);
-                let typeParameter = getTypeParameterFromMappedType(type);
+                const typeParameter = getTypeParameterFromMappedType(type);
                 // If the mapped type isn't `keyof` constraint-declared, _but_ still has modifiers preserved, and its naive instantiation won't preserve modifiers because its constraint isn't `keyof` constrained, we have work to do
                 const needsModifierPreservingWrapper = !isMappedTypeWithKeyofConstraintDeclaration(type)
                     && !(getModifiersTypeFromMappedType(type).flags & TypeFlags.Unknown)
@@ -6653,14 +6653,12 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     // We do this to ensure we retain the toplevel keyof-ness of the type which may be lost due to keyof distribution during `getConstraintTypeFromMappedType`
                     if (isHomomorphicMappedTypeWithNonHomomorphicInstantiation(type) && context.flags & NodeBuilderFlags.GenerateNamesForShadowedTypeParams) {
                         const newConstraintParam = createTypeParameter(createSymbol(SymbolFlags.TypeParameter, "T" as __String));
-                        const newTypeParam = createTypeParameter(typeParameter.symbol);
                         const name = typeParameterToName(newConstraintParam, context);
                         const target = type.target as MappedType;
-                        typeParameter = newTypeParam;
                         newTypeVariable = factory.createTypeReferenceNode(name);
                         templateType = instantiateType(
                             getTemplateTypeFromMappedType(target),
-                            makeArrayTypeMapper([getTypeParameterFromMappedType(target), getModifiersTypeFromMappedType(target)], [newTypeParam, newConstraintParam])
+                            makeArrayTypeMapper([getTypeParameterFromMappedType(target), getModifiersTypeFromMappedType(target)], [typeParameter, newConstraintParam])
                         );
                     }
                     appropriateConstraintTypeNode = factory.createTypeOperatorNode(SyntaxKind.KeyOfKeyword, newTypeVariable || typeToTypeNodeHelper(getModifiersTypeFromMappedType(type), context));

```
</details>

but the `typeParameterToName` in `typeParameterToDeclarationWithConstraint` still uses `NodeBuilderFlags.GenerateNamesForShadowedTypeParams`. Even if I save those flags on the side and generate this without this generation for shadowed type params then I get the same problem with `templateTypeNode`. 

Am I even on the right track here? Or did I misunderstand your suggestion? It might be worthwhile to take a look at the diff [here](https://github.com/microsoft/TypeScript/pull/54759/commits/eb0e1e0de45bdad1e0ed950921ae72d31f7d937d) to see how it behaves with the latest patch that I pushed out.

Reply:
> Or did I misunderstand your suggestion?

No, that diff is it, it's just unfortunate that we're manufacturing new type identities basically to work around a shortcoming in scope tracking in the type parameter name serializer. I *think* the issue as-is that the type parameter isn't "visible" to the serializer because the `context.enclosingDeclaration` is too broad - maybe if, in addition to the patch you linked, you also temporarily override `context.enclosingDeclaration` with a fake scope, as in `signatureToSignatureDeclarationHelper` it'd work. In fact, mapped types are basically signature scopes (with only a type parameter) - that's almost certainly what's missing, and what manufacturing a new type is just a kludge covering.

Reply:
To get it right:
- I should stop reassigning `typeParameter` here (like in the short `git diff` posted above in the hidden details)
- **and** I should reuse the fake scope strategy used by `signatureToSignatureDeclarationHelper`

The second thing... doesn't look completely straightforward so tbh this might take me a longer while especially since I have some more pressing reviews to address right now.